seqhash/seqhash.go

```diff
@@ -222,3 +262,181 @@ func Hash(sequence string, sequenceType SequenceType, circular bool, doubleStran
+// Used for seqhash v2
+func DecodeFlag(flag byte) (int, SequenceType, bool, bool) {
+	version := int((flag & hash2versionMask) >> hash2versionShift)
+	circularity := (flag & hash2circularityMask) != 0
+	doubleStranded := (flag & hash2doubleStrandedMask) != 0
+	dnaRnaProtein := flag & hash2typeMask
+	sequenceType := sequenceTypeByteToStringFlagMap[dnaRnaProtein]
+
+	return version, sequenceType, circularity, doubleStranded
+}
+
+// Hash2 creates a version 2 seqhash.
+func Hash2(sequence string, sequenceType SequenceType, circular bool, doubleStranded bool) ([16]byte, error) {
+	var result [16]byte
+
+	// First, get the determistic sequence of the hash
+	deterministicSequence, err := prepareDeterministicSequence(sequence, sequenceType, circular, doubleStranded)
+	if err != nil {
+		return result, err
+	}
+
+	// Build our byte flag
+	flag := EncodeFlag(2, sequenceType, circular, doubleStranded)
+	result[0] = flag
+
+	// Compute BLAKE3, then copy those to the remaining 15 bytes
+	newhash := blake3.Sum256([]byte(deterministicSequence))
+	copy(result[1:], newhash[:15])
+
+	return result, nil
+}
+
+// Hash2Fragment creates a version 2 fragment seqhash. Fragment seqhashes are
+// a special kind of seqhash that are used to identify fragments, usually
+// released by restriction enzyme digestion, rather than complete DNA
+// sequences. This is very useful for tracking genetic parts in a database: as
+// abstractions away from their container vectors, so that many fragments in
+// different vectors can be identified consistently.
+//
+// fwdOverhangLength and revOverhangLength are the lengths of both overhangs.
+// Hashed sequences are hashed with their overhangs attached. Most of the time,
+// both of these will equal 4.
+//
+// threePrimeOverhangFwd and threePrimeOverhangRev are booleans for the
+// directionality of the overhangs. The majority of restriction enzymes cut
+// leaving 5prime overhangs, so these should nearly always be false, except
+// in very special cases. threePrimeOverhangFwd replaces circular (since
+// fragments will always be linear) and threePrimeOverhangRev replaces
+// doubleStranded (since fragments will always be double stranded).
+//
+// In order to make sure fwdOverhangLength and revOverhangLength fit in the
+// hash, the hash is truncated at 13 bytes rather than 15, and both uint8 are
+// inserted. So the bytes would be:
+//
+//	flag + fwdOverhangLength + revOverhangLength + [13]byte(hash)
+//
+// 13 bytes is considered enough, because the number of fragments is limited
+// by our ability to physically produce them, while other other sequence types
+// can be found in nature.
+func Hash2Fragment(sequence string, fwdOverhangLength uint8, revOverhangLength uint8, fwdThreePrimeOverhang bool, revThreePrimeOverhang bool) ([16]byte, error) {
+	var result [16]byte
+
+	// First, run checks and get the determistic sequence of the hash
+	for _, char := range sequence {
+		if !strings.Contains("ATUGCYRSWKMBDHVNZ", string(char)) {
```

Comment:
I think that it could be better to use a map of runes instead of checking against the full string, something like:
```go
// In a global level:
var allowedChars = map[rune]bool{'A': true, 'T': true, 'U': true, 'G': true, 'C': true, 'Y': true, 'R': true, 'S': true, 'W': true, 'K': true, 'M': true, 'B': true, 'D': true, 'H': true, 'V': true, 'N': true, 'Z': true}
...
func Hash2Fragment(input string) error {
        var result [16]byte
	for _, char := range input {
		if !allowedChars[char] {
			return result, errors.New("Only letters ATUGCYRSWKMBDHVNZ are allowed for DNA/RNA. Got letter: " + string(char))
		}
	}
	return nil
}
```

Original approach:
```
% go test -benchmem -run=^$ -count=5 -bench ^BenchmarkHash2Fragment$ github.com/matiasinsaurralde/playground
goos: darwin
goarch: arm64
pkg: github.com/matiasinsaurralde/playground
BenchmarkHash2Fragment-8   	 8075786	       132.7 ns/op	      16 B/op	       1 allocs/op
BenchmarkHash2Fragment-8   	 9061473	       132.6 ns/op	      16 B/op	       1 allocs/op
BenchmarkHash2Fragment-8   	 9036169	       132.4 ns/op	      16 B/op	       1 allocs/op
BenchmarkHash2Fragment-8   	 9054006	       132.9 ns/op	      16 B/op	       1 allocs/op
BenchmarkHash2Fragment-8   	 9025446	       132.9 ns/op	      16 B/op	       1 allocs/op
```

Suggested approach:
```
% go test -benchmem -run=^$ -count=5 -bench ^BenchmarkHash2Fragment$ github.com/matiasinsaurralde/playground
goos: darwin
goarch: arm64
pkg: github.com/matiasinsaurralde/playground
BenchmarkHash2Fragment-8   	11071206	       108.5 ns/op	      16 B/op	       1 allocs/op
BenchmarkHash2Fragment-8   	11002230	       108.6 ns/op	      16 B/op	       1 allocs/op
BenchmarkHash2Fragment-8   	10999809	       108.6 ns/op	      16 B/op	       1 allocs/op
BenchmarkHash2Fragment-8   	10894378	       108.6 ns/op	      16 B/op	       1 allocs/op
BenchmarkHash2Fragment-8   	11096571	       109.0 ns/op	      16 B/op	       1 allocs/op
```

Reply:
I'm reasonably convinced. There are other places in seqhash that do something similar. Can you PR into this branch with your improvements?

Reply:
@Koeng101 added here: https://github.com/TimothyStiles/poly/pull/402 👍